### PR TITLE
Task/rich/kep 1093 Implement Volatile TTL Eviction Policy

### PR DESCRIPTION
**Pull request by rnistuk**

### CODE REVIEW

File: crud/crud_base.hpp [hunk 1]
```diff
@@ -16,7 +16,9 @@
 
 #include <include/bluzelle.hpp>
 #include <node/session_base.hpp>
+#include <policy/eviction_policy_base.hpp>
```
Comment (ebruck) [inline crud/crud_base.hpp]: These pollute the interface class and should be moved to the implementation file instead.

File: crud/crud_base.hpp [hunk 2]
```diff
@@ -16,7 +16,9 @@
...
 #include <proto/database.pb.h>
+#include <storage/storage_base.hpp>
```
Comment (ebruck) [inline crud/crud_base.hpp]: same

File: crud/crud.cpp [hunk 3]
```diff
@@ -1335,46 +1344,28 @@ crud::get_swarm_storage_usage()
...
+        if (keys_to_evict.empty())
         {
-            const size_t evicted_size = evict_key(request.header().db_uuid(), key_to_evict);
-            storage_to_free -= evicted_size < storage_to_free ? evicted_size : storage_to_free;
-            keys_to_evict.emplace_back(key_to_evict);
+            return false;
         }
+
+        std::for_each( keys_to_evict.begin(), keys_to_evict.end()
+                , [&](const auto& key)
```
Comment (ebruck) [inline crud/crud.cpp]: put , on the line above and indent lambda 4 spaces

File: policy/eviction_policy_base.hpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+
+#include <include/bluzelle.hpp>
+#include <proto/database.pb.h>
+#include <set>
```
Comment (ebruck) [inline policy/eviction_policy_base.hpp]: header order doesn't follow std (put set at end)

File: policy/eviction_policy_base.hpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#include <storage/storage_base.hpp>
+
+
+namespace bzn::policy
+{
+
+    class eviction_policy_base
```
Comment (ebruck) [inline policy/eviction_policy_base.hpp]: we are inside the policy namespace... policy is redundant. I'd just use eviction_base

File: policy/eviction_policy_base.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    {
+    public:
+        eviction_policy_base(std::shared_ptr<bzn::storage_base> storage) : storage{std::move(storage)} {}
+
+        virtual ~eviction_policy_base() = default;
+
+        virtual std::set<bzn::key_t> keys_to_evict(const database_msg &request, size_t max_size) = 0;
```
Comment (ebruck) [inline policy/eviction_policy_base.hpp]: refs to the left

File: policy/random.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+namespace bzn::policy
+{
+    class random : public eviction_policy_base
+    {
+    public:
+        random(std::shared_ptr<bzn::storage_base> storage) : eviction_policy_base{storage}
+        {
+        }
+
+        std::set<bzn::key_t> keys_to_evict(const database_msg &request, size_t max_size);
```
Comment (ebruck) [inline policy/random.hpp]: refs to the left

File: policy/random.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            : request.create().key().size() + request.create().value().size()
+        };
+
+        const bzn::key_t IGNORE_KEY{
+            request.has_update()
+            ? request.update().key()
+            : ""
+        };
+
+        std::vector<bzn::key_t> keys_to_evict;
```
Comment (ebruck) [inline policy/random.cpp]: move closer to where it's used

File: policy/random.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            // As we randomly select a key from the keys in the user's database, we *move* them from the vector of keys
+            // in the users' database to the vector of keys that need to be deleted. In the case of randomly selected
+            // the key that is being updated, we simnply remove that key from the vector of users' keys without putting
+            // it in the keys to delete.
+            const boost::random::uniform_int_distribution<> dist(0, available_keys.size() - 1);
+            const auto it = std::next(available_keys.begin(), dist(mt));
+            if (*it != IGNORE_KEY)
+            {
+                if (const auto evicted_size = this->storage->get_key_size(request.header().db_uuid(), *it))
```
Comment (ebruck) [inline policy/random.cpp]: I'd protect against an empty optional here... just to be safe

File: policy/random.cpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <policy/random.hpp>
+#include <proto/database.pb.h>
+#include <boost/src/boost/include/boost/random/mersenne_twister.hpp>
+#include <boost/src/boost/include/boost/random/uniform_int_distribution.hpp>
+
+namespace bzn::policy
+{
+    std::set<std::string>
```
Comment (ebruck) [inline policy/random.cpp]: use bzn::key_t

File: policy/volatile_ttl.hpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        volatile_ttl(std::shared_ptr<bzn::storage_base> storage) : eviction_policy_base(storage)
+        {
+        }
+
+        std::set<bzn::key_t> keys_to_evict(const database_msg &request, size_t max_size) override;
+
+    private:
+        std::vector<std::string> filter_ttls(const std::vector<std::string>& ttls, const bzn::uuid_t& db_uuid, const bzn::key_t& ignore_key);
+
+        std::vector<ttl_item> get_ttl_items(const std::string &db_uuid, std::vector<std::string> &our_ttls) const;
```
Comment (ebruck) [inline policy/volatile_ttl.hpp]: refs to the left

File: policy/volatile_ttl.hpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    struct ttl_item;
+
+    class volatile_ttl : public eviction_policy_base
+    {
+    public:
+        volatile_ttl(std::shared_ptr<bzn::storage_base> storage) : eviction_policy_base(storage)
+        {
+        }
+
+        std::set<bzn::key_t> keys_to_evict(const database_msg &request, size_t max_size) override;
```
Comment (ebruck) [inline policy/volatile_ttl.hpp]: refs to the left

File: policy/volatile_ttl.cpp [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+namespace bzn::policy
+{
+    struct ttl_item {std::string key; size_t value_size; size_t ttl;};
+
+    std::set<bzn::key_t>
+    volatile_ttl::keys_to_evict(const database_msg &request, size_t max_db_size)
+    {
+        std::set<bzn::key_t> keys_to_evict;
+
+        auto db_uuid{request.header().db_uuid()};
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: const?

File: policy/volatile_ttl.cpp [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        auto key_to_ignore{request.has_update() ? request.update().key() : "" };
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: const?

File: policy/volatile_ttl.cpp [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+namespace
+{
+    // TODO: This is a duplicate of code in crud, it would be a good idea to find a way to dry this up.
+    const std::string TTL_UUID{"TTL"};
+}
+
+
+namespace bzn::policy
+{
+    struct ttl_item {std::string key; size_t value_size; size_t ttl;};
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: bzn::key_t

File: policy/volatile_ttl.cpp [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+    std::set<bzn::key_t>
+    volatile_ttl::keys_to_evict(const database_msg &request, size_t max_db_size)
+    {
+        std::set<bzn::key_t> keys_to_evict;
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: I'd move this closer to where it's used.

File: policy/volatile_ttl.cpp [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+        // get all the TTL's in storage
+        auto all_ttls{this->storage->get_keys(TTL_UUID)};
+
+        // get the TTL's associated with our database
+        // filter out the ones for this database, if update, ignore current key
+        auto our_ttls{this->filter_ttls(all_ttls, db_uuid, key_to_ignore)};
+        if (our_ttls.empty())
+        {
+            return keys_to_evict;
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: probably could just return {}?

File: policy/volatile_ttl.cpp [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        // Create a vector of [key, value size, ttl]
+        std::vector<ttl_item> ttl_items{get_ttl_items(db_uuid, our_ttls)};
+
+        // 3: sort the keys in order of increasing ttl, secondarily sort on decreasing
+        // key value size
+        std::sort(
+                ttl_items.begin(), ttl_items.end()
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: move ttl* up to start of line and lambda on it's own indented 4 spaces

File: policy/volatile_ttl.cpp [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+
+    std::vector<ttl_item>
+    volatile_ttl::get_ttl_items(const std::string &db_uuid, std::vector<std::string> &our_ttls) const
+    {
+        std::vector<ttl_item> ttl_items;
+        auto char_reader{Json::CharReaderBuilder().newCharReader()};
+        std::transform(our_ttls.begin(), our_ttls.end(), std::back_inserter(ttl_items),
+                [&](const auto& ttl)->ttl_item
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: 4 space indent for lambda

File: policy/volatile_ttl.cpp [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+                    {
+                        bool s{(root["key"] != ignore_key) && (root["uuid"].asString() == db_uuid)};
+
+                        return s;
+                    }
+                    LOG(warning) << "Unable to parse TTL for Volatile TTL eviction: " << err;
+
+                    return false;
+                });
+        delete char_reader;
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: use unique_ptr

File: policy/volatile_ttl.cpp [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+        // 5: return the list of keys to delete
+        return keys_to_evict;
+    }
+
+
+    std::vector<ttl_item>
+    volatile_ttl::get_ttl_items(const std::string &db_uuid, std::vector<std::string> &our_ttls) const
+    {
+        std::vector<ttl_item> ttl_items;
+        auto char_reader{Json::CharReaderBuilder().newCharReader()};
```
Comment (ebruck) [inline policy/volatile_ttl.cpp]: leak?

**Review by ebruck — CHANGES_REQUESTED** [review]

File: crud/crud_base.hpp (hunk 1 above)
Comment (rnistuk) [reply crud/crud_base.hpp]: I moved the include directive to crud.hpp as policy::eviction_policy_base is used in the crud class.

File: crud/crud_base.hpp (hunk 2 above)
Comment (rnistuk) [reply crud/crud_base.hpp]: Done

File: crud/crud.cpp (hunk 3 above)
Comment (rnistuk) [reply crud/crud.cpp]: Done

File: policy/eviction_policy_base.hpp (hunk 4 above)
Comment (rnistuk) [reply policy/eviction_policy_base.hpp]: Fixed

File: policy/eviction_policy_base.hpp (hunk 5 above)
Comment (rnistuk) [reply policy/eviction_policy_base.hpp]: Refactored

File: policy/volatile_ttl.cpp (hunk 17 above)
Comment (rnistuk) [reply policy/volatile_ttl.cpp]: done

File: policy/volatile_ttl.cpp (hunk 18 above)
Comment (rnistuk) [reply policy/volatile_ttl.cpp]: Done

File: policy/volatile_ttl.cpp (hunk 19 above)
Comment (rnistuk) [reply policy/volatile_ttl.cpp]: done

File: policy/volatile_ttl.cpp (hunk 20 above)
Comment (rnistuk) [reply policy/volatile_ttl.cpp]: Made the change, also found the same bad pattern elsewhere.

File: policy/volatile_ttl.cpp (hunk 21 above)
Comment (rnistuk) [reply policy/volatile_ttl.cpp]: Not anymore, did the unique_ptr change

**Review by ebruck — APPROVED** [review]

**Comment by rnistuk** [comment]
Thank you Ed!

**Pull request merged**